shapely/ops.py

```diff
@@ -391,9 +391,9 @@ def shared_paths(g1, g2):
         The second geometry
     """
     if not isinstance(g1, LineString):
-        raise TypeError("First geometry must be a LineString")
+        raise GeometryTypeError("First geometry must be a LineString")
```

Comment:
Same here about GEOSException

Reply:
these are exceptions to the GIS methods (i.e. linear referencing is based on linear geometries, etc.), so I think they are an appropriate error type. (But yes, these are generally GEOS exceptions too).

Reply:
@mwtoews what I mean is that in PyGEOS, we let GEOS handle this case and bubble up the exception from GEOS as a GEOSException. If we are fine with doing that (xref #991), it doesn't necessarily make sense to change this error now to GeometryTypeError?

Reply:
@jorisvandenbossche I'm not fine with doing that in this case. If a call to a GEOS function is always going to fail due to an unusable type of geometry, the GEOS function shouldn't be called. An exception should be raised immediately instead. And I think an exception that's more specific than GEOSException should be raised. And I do see this pattern also used in PyGEOS, like at https://github.com/pygeos/pygeos/blob/63b474081352f4eab362a3b48e48c23328fcb3bd/pygeos/_geometry.pyx#L74, where a TypeError is raised instead of calling a GEOS geometry constructor that can't succeed.

Reply:
Hmm, this seems to be a long-running discussion from #991 and other threads. To summarize my understanding of the discussion, a key question is whether we should let GEOS errors happen and bubble these back up to the user. However, it sounded like we weren't necessarily clear on whether different types of GEOS errors would be converted into different Python error classes, thus allowing for differential handling on the user side. Barring doing that, it makes more sense to me to handle errors in Shapely code before sending over to GEOS.